### PR TITLE
feat: Add label_file support to service

### PR DESCRIPTION
**Pull request by idsulik**
Issue: https://github.com/docker/compose/issues/12311

### CODE REVIEW

**Comment by idsulik** [comment]
@ndeloof hi! What do you think about the PR, should I continue implementing the feature like I do(it uses the same way as we load env_file) or should I do it in a different way?

**Comment by ndeloof** [comment]
Sounds good to me. I'd prefer we can rely on docker/cli parser for this file so we don't have 2 implementations, but this would mean we resolve labels in docker/compose, and on the other hand format is pretty minimalist, so let's go this way

**Review by ndeloof — COMMENTED** [review]
LGTM

in this PR, you're anticipating the need to set a label_file as `required`. Do you have a use-case in mind ? We introduced this for `env_file` as a workaround for corner-case usages, but I'd prefer we don't make things unnecessary complicated if we don't have to, at least for a first iteration.

**Comment by idsulik** [comment]
> Do you have a use-case in mind

I did it to follow the current logic for the env_file. removed it + added tests

File: schema/compose-spec.json [hunk 1]
```diff
@@ -866,7 +867,34 @@
...
+            ]
+          }
+        }
+      ]
+    },
+
+    "label_file": {
+      "oneOf": [
+        {"type": "string"},
+        {
```
Comment (ndeloof) [inline schema/compose-spec.json:888]: Sounds to me this is a premature feature to add support for `required` here. `label_file` should be declared at most a `string|array of string` for user convenience

Comment (idsulik) [reply schema/compose-spec.json:888]: @ndeloof agree, fixed

File: transform/labelfile.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+		for i, e := range v {
+			v[i] = transformLabelFileValue(e)
+		}
+		return v, nil
+	default:
+		return nil, fmt.Errorf("%s: invalid type %T for label_file", p, v)
+	}
+}
+
+func transformLabelFileValue(data any) any {
```
Comment (ndeloof) [inline transform/labelfile.go:41]: not relevant anymore

File: loader/full-struct_test.go [hunk 3]
```diff
@@ -217,10 +217,24 @@ func services(workingDir, homeDir string) types.Services {
 			Ipc:      "host",
 			Uts:      "host",
 			Labels: map[string]string{
+				"FOO":                     "foo_from_label_file",
```
Comment (ndeloof) [inline loader/full-struct_test.go:220]: FYI this huge test we inherited from legacy compose parser in docker/cli.
As it is painful to maintain, we prefer to have individual test checking specific attribute is well supported (which you also provided 🥳)

Comment (idsulik) [reply loader/full-struct_test.go:220]: should I revert these changes or keep them?

**Comment by idsulik** [comment]
@ndeloof pushed changes, now label_file has type `[]string`

File: types/labelfile.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+   limitations under the License.
+*/
+
+package types
+
+import (
+	"encoding/json"
+)
+
+type LabelFile struct {
```
Comment (ndeloof) [inline types/labelfile.go:23]: unused ?

Comment (idsulik) [reply types/labelfile.go]: yes, thank you, removed

File: schema/compose-spec.json [hunk 5]
```diff
@@ -866,9 +867,6 @@
                   "path": {
                     "type": "string"
                   },
-                  "format": {
-                    "type": "string"
-                  },
```
[the remark this answers is not in the document]
Comment (ndeloof) [reply schema/compose-spec.json:871]: sure, but here you removed `format` attribute from `env_file` defintion

[the remark this answers is not in the document]
Comment (idsulik) [reply schema/compose-spec.json]: oh, accidentally deleted it, put it back

**Review by ndeloof — APPROVED** [review]

**Pull request merged**